docs/adrs/rework-batch-analysis-architecture.md

```diff
@@ -0,0 +1,82 @@
+# Rework Batch Analysis Architecture
+
+Date: 2020-09-28
+
+The current implemetnation of Batch Analysis at KBase has the following issues:
+
+* Current UI is not adequate: Users shouldn’t have to code in order to run batch analysis. Also it’s difficult to do so, even for those familiar with KBase code (have to find object names)
+* Dependency on KB Parallel: any changes to KB parallel could affect KB Batch and subsequently all other apps. 
+* Queue deadlocking: users have a max of 10 slots in the queue, with the current implementation one is taken up just to manage the jobs. Could lead to deadlock scenarios
+* Missing the ability to be able to run, manage and track jobs and their subjobs 
+* No good way to test and hard to benchmark or measure performance 
+* Code is split more than is necessary
+
+## Author(s)
+
+@leia-sefkin
+
+## Status
+
+Pending
+
+## Alternatives Considered
+
+* Writing a new Batch App
+* Implementing a Dynamic Service
+* Implementing a standalone Narrative Widget
+* Implementing a new (set) of EE2 endpoint(s)
+
+## Decision Outcome
+
+After much [discussion](https://docs.google.com/document/d/1PoiOas-hqgHONNzmCVupjmNCyvUvcned5YfZf-5CxsI/edit#) [and research](https://docs.google.com/spreadsheets/d/1FGecELaEBAWQ7ljXsR29h5RdtmYV_jMyLSiJAWcm4mE/edit#gid=492998772) the decision was to create a new endpoint (or set of endpoints) in EE2 for the purpose of running batch analysis workflows at KBase. 
+
+## Consequences
+
+We will have to implement a new narrative UI, however this was work that would happen regardless due as we are looking to improve the UX for batch upload and analysis at KBase. 
+
+As far as we have discussed, this decision only seems to have positive consequences. We will be able to streamline the code structure at KBase, make batch analysis easier to test and improve usability and job management. 
+
+Initial architecture sketches for what this could look like have been started [here](https://miro.com/app/board/o9J_kmb4y4Q=/?moveToWidget=3074457350208108487&cot=12)
+
+Still to be determined (not in scope of this ADR): 
+* What other endpoints we may need such as cancel batch
+* Other implementation details such as will the Narrative poll EE2 for updates on jobs or subscribe to Kafka. 
+
+## Pros and Cons of the Alternatives
+
+### Writing a new Batch App
+
+* `+` We have a potential first pass at implementation [here](https://github.com/bio-boris/simplebatch)
+* `+` Removes dependency on KB Parallel
+* `+` No longer uses up a slot in the queue
+* `+` Offers some improvement upon job management
+* `-` Code is still split between locations
+* `-` No good way to test and hard to benchmark or measure performance 
+* `-` Tracking and managing sub jobs is still a challenge
```

Comment:
I think there are open questions about how we would deal with creating sets of objects (another open question: is creating sets of objects in scope? I think it is, but I've been wrong before) without having a 'parent' app taking up a slot and waiting for children to finish.

Reply:
I think creating sets of objects is in scope, and we will need to figure out a way to (a) do that at the end of a batch run and (b) what to do about sets if child tasks fail.

I don't think that's necessarily in scope of this ADR, though. I felt the scope was the high level concept of just choosing to do all of this in the context of new endpoints in EE2.

Reply:
The reason I'm bringing it up here is that if we need a reduce type step as the end of the process, that affects the pros and cons list for apps. It's not clear any more that you can use an app without taking a slot, which is listed as a pro here. With sets, that may not be true.

Reply:
@MrCreosote If we need a reduce step, that affects the cons list for apps by adding requires TWO slots instead of one slot. And there's a difference between a reduce step slot, and a slot that sits there and doesn't do much at all. The reduce step slot would actually be useful
I agree the scope was just the high level concept

Reply:
I'll add some details about all of this in the consequences section

Reply:
Added more detail in consequences section and here as well. Please let me know what you think @MrCreosote and @bio-boris